components/brave_wallet/browser/zcash/rust/lib.rs

```diff
@@ -541,34 +996,626 @@ fn batch_decode(
+        Position::from(start_position),
+        scan_result.0.commitments.clone().into_iter());
+
+    if batch_insert_result.is_err() {
+        return false;
+    }
+
+    true
+}
+
+impl From<&[MerkleHashOrchard]> for MarkleHashVec {
+    fn from(item: &[MerkleHashOrchard]) -> Self {
+        let mut result : Vec<MerkleHashOrchard> = vec![];
+        for elem in item {
+            result.push(*elem);
+        }
+        MarkleHashVec(result)
+    }
+}
+
+impl CxxOrchardShardTreeBundle {
+    fn insert_commitments(self: &mut CxxOrchardShardTreeBundle,
+                          scan_result: &mut CxxBatchOrchardDecodeBundle) -> bool {
+        insert_commitments::<TREE_HEIGHT, CHUNK_SIZE, SHARD_HEIGHT>(&mut self.0.tree, scan_result)
+    }
+
+    fn calculate_witness(self: &mut CxxOrchardShardTreeBundle,
+                         commitment_tree_position: u32,
+                         checkpoint: u32) -> Box<CxxOrchardWitnessBundleResult> {
+        match self.0.tree.witness_at_checkpoint_id_caching((
+                commitment_tree_position as u64).into(), &checkpoint.into()) {
+            Ok(witness) => Box::new(CxxOrchardWitnessBundleResult::from(
+                Ok(CxxOrchardWitnessBundleValue { path: witness.path_elems().into() }))),
+            Err(_e) => Box::new(CxxOrchardWitnessBundleResult::from(Err(Error::WitnessError)))
+        }
+    }
+
+    fn truncate(self: &mut CxxOrchardShardTreeBundle, checkpoint: u32) -> bool {
+        self.0.tree.truncate_removing_checkpoint(&BlockHeight::from_u32(checkpoint)).is_ok()
+    }
+}
+
+impl CxxOrchardTestingShardTreeBundle {
+    fn insert_commitments(self: &mut CxxOrchardTestingShardTreeBundle, scan_result: &mut CxxBatchOrchardDecodeBundle) -> bool {
+        insert_commitments::<TESTING_TREE_HEIGHT, TESTING_CHUNK_SIZE, TESTING_SHARD_HEIGHT>(&mut self.0.tree, scan_result)
+    }
+
+    fn calculate_witness(self: &mut CxxOrchardTestingShardTreeBundle,
+            commitment_tree_position: u32, checkpoint: u32) -> Box<CxxOrchardWitnessBundleResult> {
+        match self.0.tree.witness_at_checkpoint_id_caching((commitment_tree_position as u64).into(), &checkpoint.into()) {
+            Ok(witness) => Box::new(CxxOrchardWitnessBundleResult::from(Ok(CxxOrchardWitnessBundleValue { path: witness.into() }))),
+            Err(_e) => Box::new(CxxOrchardWitnessBundleResult::from(Err(Error::WitnessError)))
+        }
+    }
+
+    fn truncate(self: &mut CxxOrchardTestingShardTreeBundle, checkpoint: u32) -> bool {
+        let result =  self.0.tree.truncate_removing_checkpoint(&BlockHeight::from_u32(checkpoint));
+        return result.is_ok() && result.unwrap();
+    }
+}
+
+impl CxxOrchardWitnessBundle {
+    fn size(self: &CxxOrchardWitnessBundle) -> usize {
+        self.0.path.0.len()
+    }
+
+    fn item(self: &CxxOrchardWitnessBundle, index: usize) -> [u8; 32] {
+        self.0.path.0[index].to_bytes()
+    }
+}
+
+pub struct CxxShardStoreImpl<H, const SHARD_HEIGHT: u8>  {
```

Comment:
I don't understand why this is `CxxShardStoreImpl` if it's not `for ShardStore`

Reply:
it seems to actually be the other way around and `ShardStore` is an implementation of `CxxShardStoreImpl` which seems very confusing and strange

Reply:
also why is this prefixed with `Cxx`? This doesn't appear to be a type that gets exposed to cpp

Reply:
there's also `type OrchardCxxShardStoreImpl = CxxShardStoreImpl<orchard::tree::MerkleHashOrchard, SHARD_HEIGHT>;` 
Maybe I'm misreading https://github.com/brave/brave-core/pull/26477/files#diff-24221df89af5168c9a53280e8694146539994af10ff8bd2bc0ede808b8a67dccR1313 ?

Reply:
ok, @bbondy explained that the code I linked to is implementing the ShardStore trait for CxxShardStoreImpl so I guess the Impl makes sense here, but this should only have `Cxx` if it's a type exposed to cpp